packages/e2e-tests/specs/block-transforms.test.js

```diff
@@ -65,10 +65,31 @@ const isAnExpectedUnhandledBlock = ( fileBase ) => {
 
 const setPostContentAndSelectBlock = async ( content ) => {
 	await setPostContent( content );
-	const blocks = await getAllBlocks();
-	const clientId = blocks[ 0 ].clientId;
-	await page.click( '.editor-post-title .editor-post-title__block' );
-	await selectBlockByClientId( clientId );
+	// A block may have complex logic, e.g., adds InnerBlocks that get selected.
+	// Our selection may be ignored if another block gets selected, we can not know if the a
+	// a block adds inner blocks or not or when that adds is going to happen,
+	// so we try to select the block three times.
+	for ( let i = 0; ; ) {
```

Comment:
Honestly speaking, it's a way too fragile to keep this as part of our e2e tests setup. It takes time to do re-tries and it still doesn't guarantee that those tests will pass.

What other alternatives do we have to both speed up those transform tests and to make them more predictable and explicit? @youknowriad any thoughts? I know you were also concerned about those tests.

Reply:
Hi @gziolo, thank you for sharing your thoughts.
The challenge here is not directly related to the transforms tests; it is related to adding a block and ensuring it is selected.
Blocks with inner blocks may get the inner block selected. That selection is a side effect that happens after many things, e.g., child blocks being dynamically generated or requested from a server. If we want to select the parent, the e2e test action of selecting the parent may be fired before the effect selects the child. If that happens, the test fails because the test expected the parent to be selected, and the selected block is the child.
I would say if we try three times at most, the effect to select the child should happen, and things should work (in practice, and these retries should be very rare).
This problem is not exclusive to these tests. In any case, we add a block with child blocks that may be selected; we may have identical fails. So, I think it would be good to have a solution; otherwise, we will not have a reliable way to test these blocks.

That said, the current reason why these tests become unstable was a real minor bug with the navigation block. Recently it got auto-populated with top-level pages, but instead of keeping the selection on the navigation block, the selection was going to the child. I'm solving this bug at https://github.com/WordPress/gutenberg/pull/17835.

So I guess even without this PR, the test will get more stability if that bug fix is merged.
But I was thinking long term, e.g., in a quote block. If it used inner blocks, probably the child paragraph should be selected after the insert so the user can type right away.
In a scenario like that where I want to write a test that adds a quote block, and then I want to change the style of the quote block, how can I do that?
I need to select the quote block, to change its style, so I need to select the quote block after the selection of the paraph happened; otherwise, I may select the quote, but then the paragraph gets selected and my test breaks because the expected style is not available.

In case there some suggestions for this general problem, it would be important in the future to ensure these cases are tested easily.

Regarding a way to make transform texts more predictable, I think one way is manual tests. But, that would mean that most of the transforms would not be exercised at all. At most, we would create eight tests, but we have many more transforms.
On the other side, these eight manual tests would probably be more specific, so we can make them more resilient, and they would probably be more effective in testing the transforms they exercice.
I think it is a tradeoff between having something that tries to find bugs everywhere but may have some fragility because it is hard to write a single test that works everywhere. And smaller tests that don't test everything but are more effective testing the subset they exercise.

The reason why the transformed test was added was the constant regressions in the transforms. So, If we remove the test we need to write some "manual" transform tests for the most popular transforms soon after the removal (I can work on that if we decide to remove this test).